### PR TITLE
fix: shorten email otp 

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
* Shortens email otp sent when the user requests for a confirmation, recovery, invite, email change, reauthentication and magic links
* Email otps will now be sent with a fixed length of 10 and will only contain alphanumeric characters
* Adds a partial unique index on the following db fields: `confirmation_token`, `recovery_token`, `email_change_token_new`, `email_change_token_current`, `reauthentication_token`

### CODE REVIEW

File: api/mail.go [hunk 1]
```diff
@@ -280,3 +306,30 @@ func (a *API) validateEmail(ctx context.Context, email string) error {
 	}
 	return nil
 }
+
+// generateUniqueEmailOtp returns a unique otp
+func generateUniqueEmailOtp(tx *storage.Connection, tokenType string) (string, error) {
```
Comment (J0) [inline api/mail.go:311]: /nit can consider changing tokenType to an enum in the future to guard against accidental typos on the `tokenType` field

**Review by J0 — APPROVED** [review]
Looks good to me

File: api/mail.go (hunk 1 above)
Comment (kangmingtay) [reply api/mail.go:311]: that's a good suggestion, i'll make this change in the same PR!

File: crypto/crypto.go [hunk 2]
```diff
@@ -37,3 +37,22 @@ func GenerateOtp(digits int) (string, error) {
...
+			return "", errors.WithMessage(err, "Error generating otp from charset")
+		}
+		b[i] = charset[val.Int64()]
+	}
+	return string(b), nil
+}
+
+// GenerateEmailOtp generates a random n-length alphanumeric otp
+func GenerateEmailOtp(length int) (string, error) {
+	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
Comment (awalias) [inline crypto/crypto.go:56]: I think it's ok but are there any situations where upper case chars could get squashed?

File: crypto/crypto.go [hunk 3]
```diff
@@ -37,3 +37,22 @@ func GenerateOtp(digits int) (string, error) {
 	otp := fmt.Sprintf(expr, val.String())
 	return otp, nil
 }
+
+// GenerateOtpFromCharset generates a random n-length otp from a charset
+func GenerateOtpFromCharset(length int, charset string) (string, error) {
+	b := make([]byte, length)
+	for i := range b {
+		val, err := rand.Int(rand.Reader, big.NewInt(int64(len(charset))))
```
Comment (awalias) [inline crypto/crypto.go:45]: is this how we did it previously? If not have we checked that it's a cryptographically secure way of doing random number generation in go?

**Review by awalias — APPROVED** [review]
looks good just a couple of questions

File: crypto/crypto.go (hunk 3 above)
Comment (awalias) [reply crypto/crypto.go:45]: ah yeah looks ok: https://pkg.go.dev/crypto/rand

**Comment by kangmingtay** [comment]
block on this for the time being because we need to run the migration to create the new indices concurrently which isn't supported by the current migration tool (soda) 

will be swapping that out for [sql-migrate](https://github.com/rubenv/sql-migrate)

**Comment by RatulSaha** [comment]
A typical user journey is to look at the OTP in phone while doing it in web/iPad. This is quite common to not context switch, as people are often comfortable checking email (like SMS) in phone.

On that note, can we have a consistent experience, where we do not assume that it will be copy pasted? Most providers do lowercase alphabets only, for example. Making them a multiple of 4 is also common for the same reason, it makes it possible to remember chunks and type it out.

**Pull request merged**